### DISCUSSION TITLE
Depending and using root.jsx loader data inside route loaders

### DISCUSSION THREAD

**Discussion by KasparRosin** [remix-run/remix#3811] [Q&A] [answered]
Have an issue where we have a API request, that returns bunch of data from the api, that all other routes can depend on.
I've tried searching how to pass loader data from root.jsx to child loaders (routes/index.js), but so far have not come up with anything.

I've tried passing the parent data through `<Outlet context={loaderData} />`, but this didn't work.
I noticed that loader props contains `context` key, which is undefined for me. How to provide context value there?

I looked up that remix suggests calling the request in every route, but doesn't this affect performance when we have deeply nested routes and each one of them requests the config/data from api.

**Comment by lili21** [comment]
You can't do this. the requests inside loader are sending parallel. it won't affect performance. 

with that in mind, if there are dependencies with the request, you should put it together, inside one loader.

**Reply by KasparRosin** [reply]
Yeah, but if the route is x3 nested, it will still be x3 fetch calls made assuming we need the dependency in all 3 routes.

**Reply by lili21** [reply]
You need to call the x3 fetch anyway right?

**Reply by lili21** [reply]
oh, I get what you mean. in that way, I guess you could put all the requests in root loader?

**Reply by KasparRosin** [reply]
Can't since some loaders depend on url params.

**Reply by KasparRosin** [reply]
example route: blogs/$user/$category

**Reply by lili21** [reply]
then how about put it on query ? like do the base fetch on root, and redirect with query? share the data through url

**Comment by jgb-solutions** [comment]
Have you tried the `useMatches` hook? It allows you to have access to loader data for every page, including root.tsx's loader. https://remix.run/docs/en/v1/api/remix#usematches

**Reply by lili21** [reply]
I think what he wants is access other loader's data in loader.

**Reply by jgb-solutions** [reply]
Oh in this case `useMatches` won't help here. it's for the frontend.

**Comment by sergiodxa** [comment, marked as answer]
If you want to share code between loaders you will have to cache it somewhere, it could be an in-memory cache with something like leu-cache or outside the process using Redis.

You can't share data between requests because of two reasons


They run in parallel, so the moment the root loader is running all the other loaders are running too, so you can't expect the root loader ro tun first, that would cause a waterfall of requests that would impact your perf more than doing the same request multiple times

They may not even run at all, on SSR all loaders run as functions together, but on a client-side navigation Remix does a fetch per loader, this means your loaders run as completely separate HTTP requests.

And additionally the parent loaders may not be fetched at all, for example the root loader will not be fetched on a client-side navigation unless you change the search params or submit an action first, this means if you only fetched on root, your others loaders wouldn't have access to that data on a client-side navigation.

Because of the last thing, you shouldn't expect the root loader to have already fetched some data, all your loaders should fetch the data, but you can cache it and do a cache-first approach where you check if something is in cache and if not fetch it, and do this on every loader.

**Reply by jgb-solutions** [reply]
@sergiodxa your website is broken by the way. Looks like it's because of the collectednotes service certs that expired right now. Maybe you should cache those responses as well. I couldn't read the article I wanted on your site. Had to use a google cache.

**Reply by sergiodxa** [reply]
@jgb-solutions it's working for me right now, the SSL issue was already fixed a few hours ago 🤔

**Reply by jgb-solutions** [reply]
@sergiodxa that's because `https://collectednotes.com/` fixed their certs. Otherwise I wouldn't know your site uses it, right? It's all server side.

**Reply by lili21** [reply]
what will happen to the child route fetch request if a parent route does the redirect?

**Reply by sergiodxa** [reply]
> what will happen to the child route fetch request if a parent route does the redirect?

They will be ignored

**Reply by sergiodxa** [reply]
> @sergiodxa that's because `https://collectednotes.com/` fixed their certs. Otherwise I wouldn't know your site uses it, right? It's all server side.

Yes, everything happens on loaders

**Reply by KasparRosin** [reply]
Thanks for your answer! @sergiodxa 
Caching was the next idea in my book, but I usually want to avoid it when already working with an external api that caches the results.

What about the context key that is provided in the loader? `loader({context, request, params})`, where does that come from?
